hc-nested-list-behavior.html

```diff
@@ -0,0 +1,178 @@
+      itemIdName: {
+        type: String,
+        value: 'id'
+      },
+      /* The property name for the parentId on each item. */
+      itemParentIdName: {
+        type: String,
+        value: 'parentId'
+      },
+      /**
+       * The flat array of items to build the nested list structure from, requires an id and
+       *  parentId as defined by the itemIdName and itemParentIdName properties.
+       *
+       * @type {Object[]}
+       */
+      items: Array,
+      /**
+       * A nested representation of items, maintains list state; gets rebuilt when items changes,
+       *  when lazy-loading data add a new key-value pair to this object containing the data.  The
+       *  key value is used for looking up parentId to fetch the list of child items.  All item data
+       *  is taken from the items array.
+       */
+      parts: {
+        type: Object,
+        notify: true
+      }
+    },
+    observers: ['_computeParts(items)'],
+    listeners: {
+      toggle: 'toggleOpen'
+    },
+    /**
+     * Takes a parts object and optionally a start point for an array then iterates over the
+     *  object starting at the root if there isn't a location already set and finds all open nodes
+     *  recursively calling itself to ensure order is preserved once flattened.
+     *
+     * @param {Object} parts
+     * @param {Object[]} [items=parts.root]
+     * @param {String} [idName=this.itemIdName]
+     * @returns {Object[]}
+     */
+    _computeItems(parts, items = parts.root, idName = this.itemIdName) {
+      // Can't calculate items without parts or the name of the id property.
+      if (!idName || !parts) return [];
+
+      const newItems = [];
+
+      // Push all items into the new array, and if there is a child item then push all of the children.
+      items.forEach(item => {
+        item.open = Boolean(item.open);
+        newItems.push(item);
+        if (item.open)
+          newItems.push(...this._computeItems(parts, parts[item[idName]] || []));
+      });
+
+      return newItems;
+    },
+    /**
+     * Takes the flat items structure and builds out a map of parent-child elements which gets
+     *  stored and used for optimized rendering later.
+     *
+     * @param {Object[]} items
+     */
+    _computeParts(items) {
+      const parts = {};
+
+      items.forEach(item => {
+        // Fetch the parentId or set it to 'root' if there isn't one
+        const parentId = this._getParentId(item);
+        if (!parts[parentId]) parts[parentId] = [];
+        parts[parentId].push(item);
+      });
+
+      this.set('parts', parts);
+    },
+    /**
+     * Used to fetch the parentId if it's available or default back to 'root'.
+     *
+     * @param {Object} item
+     * @param {String} [parentIdName=this.itemParentIdName]
+     * @returns {String|Number}
+     */
+    _getParentId(item, parentIdName = this.itemParentIdName) {
+      return (Boolean(item[parentIdName]) || item[parentIdName] === 0) ?
```

Comment:
The returned value shouldn't be any of the falsey values: false, '', null, undefined, NaN; except for 0 is valid.  The first part being cast to a Boolean isn't really required though, just something I forgot to remove.